cassiopeia/data/CassiopeiaTree.py

```diff
@@ -1144,6 +1168,37 @@ def leaves_in_subtree(self, node) -> List[str]:
 
         return self.__cache["subtree"][node]
 
+    def subset_tree_at_node(
+        self, node: str, copy: bool = False
+    ) -> Optional["CassiopeiaTree"]:
+        """Subset CassiopeiaTree object at node.
+
+        Given a node in the CassiopeiaTree, subset the entire tree object
+        to only the nodes that fall below that node.
+
+        Args:
+            node: Node identifier in the object.
+            copy: Return a copy or operate in place.
+        
+        Returns:
+            A subset CassiopeiaTree object if copy is set to true, else None. 
+        """
+        if copy:
+            new_tree = self.copy()
+            new_tree.subset_tree_at_node(node)
+            return new_tree
+
+        nodes_in_subtree = self.depth_first_traverse_nodes(source=node)
+        to_remove = list(set(self.nodes) - set(nodes_in_subtree))
+        for node in to_remove:
+            self.__remove_node(node)
+
+        # reset cache because we've changed the tree topology
+        self.__cache = {}
+
+        # Remove all removed nodes from data fields
+        self.__register_data_with_tree()
```

Comment:
Why don't we make `__register_data_with_tree` clear the cache? Is there ever a time one would `__register_data_with_tree` and not clear the cache? Sounds like a source of bugs. The other times `__register_data_with_tree ` is used is in `add_leaf` and `remove_leaf_and_prune_lineage`, where the same pattern as in this PR is used. This leads to a duplicated pattern of `self.__cache = {}` followed by `self.__register_data_with_tree()`.

Reply:
I agree -- I've changed the logic there.